### PR TITLE
feature/#51 crud for books

### PR DESCRIPTION
**Pull request by Kerosz**
**__Task Description__**
CRUD Operations for books

**__Technical Considerations__**
- [x] Create Book Schema/Entity
- [x] Create Book Controller
- [x] Create Book API Properties
- [x] Validate DTOs

**__Additional Considerations__**
[https://app.zenhub.com/files/401346050/2ef2f170-434b-4e74-b242-f29bd91a828a/download](https://app.zenhub.com/files/401346050/2ef2f170-434b-4e74-b242-f29bd91a828a/download)

Resolves #51

### CODE REVIEW

File: src/book/dto/create-book.dto.ts [hunk 1]
```diff
@@ -50,8 +50,8 @@ export class CreateBookDto {
   @IsDefined({ message: "Categories must not be empty!" })
   categories!: string[];
 
-  @ApiProperty({ description: "Date of the book release", default: new Date(Date.now()) })
-  @IsDate({ message: "Realease Date must be a valid Date type!" })
+  @ApiProperty({ description: "Date of the book release", default: 1519211809934 })
```
Comment (ArturoGrullon) [inline src/book/dto/create-book.dto.ts:53]: Why in stead of date.now() that string?

File: src/utils/file-system.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return await fs.writeFile(path, data);
+  } catch (error: unknown) {
+    if (error instanceof Error) {
+      console.error(`Got an error trying to read the file: ${error.message}`);
+    } else {
+      console.error(error);
+    }
+  }
+}
+
```
Comment (ArturoGrullon) [inline src/utils/file-system.ts:37]: What's the reason behind this file, why would we need to read, write into a file?

File: src/book/book.service.ts [hunk 3]
```diff
@@ -1,28 +1,113 @@
...
+    });
+    const pPdf: ProcessFile = new ProcessFile({
+      fileBuffer: pdfFile,
+      fileType: "file",
+      customExt: "pdf",
+      subDir: formattedBookName,
+    });
+
+    const coverPath: string = await pImg.resizeAndSave({ width: 432, height: 574 });
+    const bookPath: string = await pPdf.saveFile();
```
Comment (ArturoGrullon) [inline src/book/book.service.ts:31]: Any particular reason behind this values?

**Review by ArturoGrullon — APPROVED** [review]
Over all everything seems pretty organized, yet I'm not very familiar with the nest framework, as well, libraries used to give a proper review. 

I did notice the reading/writing of local files, I'm assuming it's something related with the optimization of images. 

I type some comments, nothing bad, just confused.

File: src/book/book.service.ts (hunk 3 above)
Comment (Kerosz) [reply src/book/book.service.ts:31]: By this value, you mean the width and hight value ? There is no reason why it is 432x574, I felt this is small enough to not cause a lot of issues, but not too small so it is still usable without scaling the cover image on the frontend.

If by the value you mean `coverPath` and `bookPath`, they are the paths on where the `PDF` and the cover `img` are saved on the server

File: src/utils/file-system.ts (hunk 2 above)
Comment (Kerosz) [reply src/utils/file-system.ts:37]: To save the cover `img` and `pdf` on the file system. This file is wrapping `fs` promises module, to catch the errors. The file is using internally in the `process-file.ts` utility

File: src/book/dto/create-book.dto.ts (hunk 1 above)
Comment (Kerosz) [reply src/book/dto/create-book.dto.ts:53]: It doesn't really matter, mainly I changed, it's the value that will be shown in the swagger API docs, I changed the `new Date` as I decided to just save an epoch date, which is a 10 digit number, I could have left the `Date.now()` though.

And the reason is a number string, is becacuse in order to upload file you need to use a `multipar/form-data` to send the request, which means everything will be a string, no matter what you send, so you have to parse it further on the server.

**Comment by Kerosz** [comment]
I'm going to assume everything is good, and squash it

**Pull request merged**